src/v/datalake/coordinator/frontend.cc

```diff
@@ -0,0 +1,246 @@
+        ::model::datalake_coordinator_nt);
+    if (!md) {
+        return std::nullopt;
+    }
+    iobuf temp;
+    write(temp, tp);
+    auto bytes = iobuf_to_bytes(temp);
+    auto partition = murmur2(bytes.data(), bytes.size())
+                     % md->get().get_configuration().partition_count;
+    return ::model::partition_id{static_cast<int32_t>(partition)};
+}
+
+ss::future<bool> frontend::ensure_topic_exists() {
+    // todo: make these configurable.
+    static constexpr int16_t default_replication_factor = 3;
+    static constexpr int32_t default_coordinator_partitions = 3;
+
+    auto& topics = _topics_frontend->local();
+    const auto& metadata = topics.get_metadata_cache();
+
+    if (metadata.get_topic_metadata_ref(::model::datalake_coordinator_nt)) {
+        co_return true;
+    }
+    auto replication_factor = default_replication_factor;
+    if (replication_factor > (int16_t)metadata.node_count()) {
+        replication_factor = 1;
+    }
+
+    cluster::topic_configuration topic{
+      ::model::datalake_coordinator_nt.ns,
+      ::model::datalake_coordinator_nt.tp,
+      default_coordinator_partitions,
+      replication_factor};
+
+    // todo: fix this by implementing on demand raft
+    // snapshots.
+    topic.properties.cleanup_policy_bitflags
+      = ::model::cleanup_policy_bitflags::none;
+
+    try {
+        auto res = co_await topics.autocreate_topics(
+          {std::move(topic)},
+          config::shard_local_cfg().create_topic_timeout_ms());
+        vassert(res.size() == 1, "expected exactly one result");
+        if (
+          res[0].ec != cluster::errc::success
+          && res[0].ec != cluster::errc::topic_already_exists) {
+            vlog(
+              datalake::datalake_log.warn,
+              "can not create topic: {} - error: {}",
+              ::model::datalake_coordinator_nt,
+              cluster::make_error_code(res[0].ec).message());
+            co_return false;
+        }
+        co_return true;
+    } catch (const std::exception_ptr& e) {
+        vlog(
+          datalake::datalake_log.warn,
+          "can not create topic {} - exception: {}",
+          ::model::datalake_coordinator_nt,
+          e);
+        co_return false;
+    }
+}
+
+ss::future<add_translated_data_files_reply>
+frontend::add_translated_data_files_locally(
+  add_translated_data_files_request request,
+  const ::model::ntp& coordinator_partition,
+  ss::shard_id shard) {
+    co_return co_await _partition_mgr->invoke_on(
+      shard,
+      [&coordinator_partition,
+       req = std::move(request)](cluster::partition_manager& mgr) mutable {
+          auto partition = mgr.get(coordinator_partition);
+          if (!partition) {
+              return ssx::now(
+                add_translated_data_files_reply{coordinator_errc::not_leader});
+          }
+          auto stm = partition->raft()->stm_manager()->get<coordinator_stm>();
+          return stm->add_translated_data_file(std::move(req));
```

Comment:
Just making sure / wondering about broader structure: long term do you expect this to call into a to-be-implemented coordinator that checks for leadership, dedupes files, etc.?